### PR TITLE
bring back transparent websocket (Refs #945)

### PR DESCRIPTION
**Pull request by melekes**
Requires https://github.com/tendermint/tmlibs/pull/96

### CODE REVIEW

File: rpc/lib/types/types.go [hunk 1]
```diff
@@ -135,10 +137,14 @@ type WSRPCConnection interface {
 	GetRemoteAddr() string
 	WriteRPCResponse(resp RPCResponse)
 	TryWriteRPCResponse(resp RPCResponse) bool
+	GetEventSubscriber() EventSubscriber
+}
 
-	AddSubscription(string, interface{}) error
-	DeleteSubscription(string) (interface{}, bool)
-	DeleteAllSubscriptions()
+// EventSubscriber mirros tendermint/tendermint/types.EventBusSubscriber
```
Comment (melekes) [inline rpc/lib/types/types.go]: or it's ok to use `tmtypes` here?

Comment (ebuchman) [reply rpc/lib/types/types.go]: hmm good question. let's keep the duplication for now and think about how we might improve package structure for things like this

File: rpc/client/httpclient.go [hunk 2]
```diff
@@ -316,34 +309,27 @@ func (w *WSEvents) redoSubscriptions() {
...
-			if ch := w.getSubscription(result.Query); ch != nil {
+			w.mtx.RLock()
+			if ch, ok := w.subscriptions[result.Query]; ok {
 				ch <- result.Data
 			}
-		case <-w.quit:
-			// send a message so we can wait for the routine to exit
-			// before cleaning up the w.ws stuff
-			w.done <- true
+			w.mtx.RUnlock()
```
Comment (melekes) [inline rpc/client/httpclient.go]: probably better to move up a little bit (i.e. unblock before writing to channel)

File: rpc/core/events.go [hunk 3]
```diff
@@ -44,20 +44,19 @@ func Subscribe(wsCtx rpctypes.WSRPCContext, query string) (*ctypes.ResultSubscri
...
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to add subscription")
+		return nil, errors.Wrap(err, "failed to parse query")
 	}
 
 	ctx, cancel := context.WithTimeout(context.Background(), subscribeTimeout)
 	defer cancel()
 	ch := make(chan interface{})
-	err = eventBus.Subscribe(ctx, addr, q, ch)
+	es := wsCtx.GetEventSubscriber()
```
Comment (melekes) [inline rpc/core/events.go]: ???? unused

File: rpc/client/httpclient.go (hunk 2 above)
Comment (melekes) [reply rpc/client/httpclient.go]: no. it is intended, so we can't close the channel

Comment (melekes) [reply rpc/client/httpclient.go]: wrote a comment

File: rpc/client/httpclient.go [hunk 4]
```diff
@@ -204,108 +204,101 @@ type WSEvents struct {
...
 	w.mtx.Lock()
-	defer w.mtx.Unlock()
-	ch, ok := w.subscriptions[query]
+	ch, ok := w.subscriptions[q]
 	if ok {
 		close(ch)
-		delete(w.subscriptions, query)
+		delete(w.subscriptions, q)
 	}
+	w.mtx.Unlock()
```
Comment (ebuchman) [inline rpc/client/httpclient.go]: Why not keep the defer? I know there's a small performance price but probably worth it for maintainability

File: rpc/client/httpclient.go [hunk 5]
```diff
@@ -316,34 +309,27 @@ func (w *WSEvents) redoSubscriptions() {
...
-		case resp := <-w.ws.ResponsesCh:
-			// res is json.RawMessage
+		case resp, ok := <-w.ws.ResponsesCh:
+			if !ok {
+				return
+			}
 			if resp.Error != nil {
-				// FIXME: better logging/handling of errors??
-				fmt.Printf("ws err: %+v\n", resp.Error.Error())
+				w.Logger.Error("WS error", "err", resp.Error.Error())
```
Comment (ebuchman) [inline rpc/client/httpclient.go]: Woo!

File: rpc/core/events.go [hunk 6]
```diff
@@ -44,20 +44,19 @@ func Subscribe(wsCtx rpctypes.WSRPCContext, query string) (*ctypes.ResultSubscri
...
+	if es == nil {
+		es = eventBus
```
Comment (ebuchman) [inline rpc/core/events.go]: :o hmm.

Seems a bit too magical.

This whole package is a bit magical with the globals. Should we look to introduce an explicit `RPCPipe` that contains all these guys, and make all these rpc funcs methods on that ?

File: rpc/core/events.go [hunk 7]
```diff
@@ -100,18 +95,31 @@ func Subscribe(wsCtx rpctypes.WSRPCContext, query string) (*ctypes.ResultSubscri
...
+	if err != nil {
+		return nil, err
+	}
 	return &ctypes.ResultUnsubscribe{}, nil
 }
+
+func eventBusFor(wsCtx rpctypes.WSRPCContext) tmtypes.EventBusSubscriber {
+	es := wsCtx.GetEventSubscriber()
+	if es == nil {
+		es = eventBus
```
Comment (ebuchman) [inline rpc/core/events.go]: :o hmm.

So I guess this happens if someone calls NewWebsocketManager and forgets to pass the option to set the eventBus? 

Also maybe its time to make all these funcs methods on a `RPCPipe` struct or something so we can do away with the globals. Agree ?

File: rpc/lib/server/handlers.go [hunk 8]
```diff
@@ -499,28 +503,6 @@ func (wsc *wsConnection) TryWriteRPCResponse(resp types.RPCResponse) bool {
 	}
 }
 
-func (wsc *wsConnection) AddSubscription(query string, data interface{}) error {
```
Comment (ebuchman) [inline rpc/lib/server/handlers.go]: good riddance!

**Pull request merged**